File: web/src/crypto/envelope.ts
/**
 * The sealed envelope, client side.
 *
 * This must match internal/pairing/envelope.go byte for byte. The wire format
 * and the nonce layout are specified in specs/001-lan-file-transfer/contracts/pairing.md;
 * anything here that drifts from the Go implementation produces an
 * authentication failure rather than a silent misinterpretation, which is the
 * one saving grace of an AEAD.
 *
 * `crypto.subtle` is deliberately not used. The mobile page is served over a
 * plain local address, which browsers never treat as a secure context, so the
 * native cryptography API does not exist there. @noble/ciphers is audited and
 * pure JavaScript, so it works in both contexts.
 */

import { chacha20poly1305 } from '@noble/ciphers/chacha';

export const PROTOCOL_VERSION = 1;

/** Which counter space a message belongs to. Separating the directions is what
 *  stops a server message being reflected back at the server. */
export const ClientToServer = 0;
export const ServerToClient = 1;

const COUNTER_SIZE = 8;
const NONCE_SIZE = 12;
const KEY_SIZE = 32;

export class ReplayError extends Error {
  constructor() {
    super('replayed or out-of-order counter');
    this.name = 'ReplayError';
  }
}

export class AuthenticationError extends Error {
  constructor() {
    super('envelope failed to authenticate');
    this.name = 'AuthenticationError';
  }
}

/** Builds the 12-byte nonce: direction, then the counter, then zero padding. */
function nonce(direction: number, counter: bigint): Uint8Array {
  const out = new Uint8Array(NONCE_SIZE);
  out[0] = direction;
  new DataView(out.buffer).setBigUint64(1, counter, false);
  return out;
}

/**
 * Builds the associated data that binds a message to its endpoint.
 *
 * Layout: method, a zero byte, path, a zero byte, then the version as a
 * big-endian uint32. Without this, an envelope captured from a harmless
 * endpoint could be replayed against a dangerous one.
 */
function associatedData(method: string, path: string, version: number): Uint8Array {
  const encoder = new TextEncoder();
  const m = encoder.encode(method);
  const p = encoder.encode(path);

  const out = new Uint8Array(m.length + 1 + p.length + 1 + 4);
  let offset = 0;
  out.set(m, offset);
  offset += m.length + 1; // the separator is already zero
  out.set(p, offset);
  offset += p.length + 1;
  new DataView(out.buffer).setUint32(offset, version, false);
  return out;
}

/** Seals and opens control payloads for one session. */
export class Envelope {
  private readonly key: Uint8Array;
  private readonly sendDirection: number;
  private readonly recvDirection: number;

  private sendCounter = 0n;
  private recvHighWater = 0n;
  private seenAny = false;

  constructor(key: Uint8Array, sendDirection: number = ClientToServer) {
    if (key.length !== KEY_SIZE) {
      throw new Error(`session key must be ${KEY_SIZE} bytes, got ${key.length}`);
    }
    this.key = key;
    this.sendDirection = sendDirection;
    this.recvDirection = sendDirection === ServerToClient ? ClientToServer : ServerToClient;
  }

  /** Encrypts a payload for one request. */
  seal(method: string, path: string, plaintext: Uint8Array): Uint8Array {
    this.sendCounter += 1n;

    const sealed = chacha20poly1305(
      this.key,
      nonce(this.sendDirection, this.sendCounter),
      associatedData(method, path, PROTOCOL_VERSION),
    ).encrypt(plaintext);

    const out = new Uint8Array(COUNTER_SIZE + sealed.length);
    new DataView(out.buffer).setBigUint64(0, this.sendCounter, false);
    out.set(sealed, COUNTER_SIZE);
    return out;
  }

  /** Decrypts a payload, refusing replays and out-of-order counters. */
  open(method: string, path: string, sealed: Uint8Array): Uint8Array {
    if (sealed.length < COUNTER_SIZE) {
      throw new AuthenticationError();
    }

    const counter = new DataView(
      sealed.buffer,
      sealed.byteOffset,
      sealed.byteLength,
    ).getBigUint64(0, false);

    // Strictly increasing. Counters start at 1, so zero is malformed.
    if (counter === 0n) throw new AuthenticationError();
    if (this.seenAny && counter <= this.recvHighWater) throw new ReplayError();

    let plaintext: Uint8Array;
    try {
      plaintext = chacha20poly1305(
        this.key,
        nonce(this.recvDirection, counter),
        associatedData(method, path, PROTOCOL_VERSION),
      ).decrypt(sealed.subarray(COUNTER_SIZE));
    } catch {
      // The high-water mark is deliberately not advanced on failure, so a
      // forged message with a huge counter cannot wedge the session.
      throw new AuthenticationError();
    }

    this.recvHighWater = counter;
    this.seenAny = true;
    return plaintext;
  }
}

/** Base64 helpers. The wire carries base64 because the payload travels in a
 *  request body that must survive whatever a browser does to it. */
export function toBase64(bytes: Uint8Array): string {
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

export function fromBase64(text: string): Uint8Array {
  const binary = atob(text.trim());
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

File: internal/pairing/envelope.go
// Package pairing implements the trust handshake, session credentials, and the
// authenticated envelope that protects control traffic.
//
// Everything here is designed on the assumption that a passive observer reads
// every byte. In simple mode the channel is plain HTTP on a shared network, and
// constitution v2.0.1 still requires pairing exchanges, credentials, and
// metadata to be encrypted in every mode.
package pairing

import (
	"encoding/binary"
	"errors"
	"fmt"
	"sync"

	"golang.org/x/crypto/chacha20poly1305"
)

// Direction separates the two counter spaces. Without it, a message the server
// sent could be replayed back at the server with a counter it would accept.
type Direction uint8

const (
	ClientToServer Direction = 0
	ServerToClient Direction = 1
)

// Wire layout: an 8-byte big-endian counter followed by the AEAD ciphertext.
// The counter travels in the clear because the receiver needs it to build the
// nonce, and it is covered by the authentication tag through the nonce itself.
const counterSize = 8

// Errors an envelope can produce. They are distinguished because the corrective
// action differs: a replay means reload the page, a bad key means re-pair.
var (
	ErrReplay         = errors.New("replayed or out-of-order counter")
	ErrMalformed      = errors.New("malformed envelope")
	ErrAuthFailed     = errors.New("authentication failed")
	ErrCounterWrapped = errors.New("counter space exhausted")
)

// Envelope seals and opens control-plane payloads for one session.
//
// It is safe for concurrent use. Sealing takes a lock because the counter must
// advance exactly once per message: two goroutines sharing a counter value
// would reuse a nonce, which is the one thing that breaks this construction
// outright.
type Envelope struct {
	mu   sync.Mutex
	aead interface {
		Seal(dst, nonce, plaintext, additionalData []byte) []byte
		Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error)
		NonceSize() int
	}

	sendDir Direction
	recvDir Direction

	sendCounter uint64
	// recvHighWater is the highest counter accepted so far. Anything at or
	// below it is a replay.
	recvHighWater uint64
	seenAny       bool
}

// NewEnvelope builds an envelope from a 32-byte session key.
//
// send is the direction this side writes in; the other direction is implied,
// so a caller cannot accidentally configure both sides to share one space.
func NewEnvelope(key []byte, send Direction) (*Envelope, error) {
	return NewEnvelopeAt(key, send, 0)
}

// NewEnvelopeAt builds an envelope whose send counter resumes at start.
//
// A browser page keeps its counter in memory, so a reload builds a new envelope
// while the peer still remembers the highest counter it accepted. Starting
// again at zero makes every message look like a replay and leaves the session
// refused until the peer restarts. The client therefore claims a fresh range of
// counters at each load, and this is how one is built. Nothing is weakened:
// counters still only ever increase, which is the property the replay check
// rests on.
func NewEnvelopeAt(key []byte, send Direction, start uint64) (*Envelope, error) {
	if len(key) != chacha20poly1305.KeySize {
		return nil, fmt.Errorf("session key must be %d bytes, got %d",
			chacha20poly1305.KeySize, len(key))
	}

	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, fmt.Errorf("build cipher: %w", err)
	}

	recv := ServerToClient
	if send == ServerToClient {
		recv = ClientToServer
	}

	return &Envelope{aead: aead, sendDir: send, recvDir: recv, sendCounter: start}, nil
}

// nonce builds the 12-byte nonce: one direction byte, the 8-byte counter, and
// three zero bytes. Direction and counter together are unique per key, which is
// what the AEAD requires.
func nonce(dir Direction, counter uint64) []byte {
	var n [chacha20poly1305.NonceSize]byte
	n[0] = byte(dir)
	binary.BigEndian.PutUint64(n[1:9], counter)
	return n[:]
}

// aad binds a message to the endpoint it was meant for.
//
// Without this, an envelope captured from one request could be replayed against
// a different path with the same counter on a fresh session. The protocol
// version is included so a downgrade cannot reuse traffic from another version.
func aad(method, path string, version int) []byte {
	out := make([]byte, 0, len(method)+len(path)+8)
	out = append(out, method...)
	out = append(out, 0)
	out = append(out, path...)
	out = append(out, 0)
	return binary.BigEndian.AppendUint32(out, uint32(version)) //nolint:gosec // version is a small constant
}

// Seal encrypts a payload for one request.
func (e *Envelope) Seal(method, path string, version int, plaintext []byte) ([]byte, error) {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.sendCounter == ^uint64(0) {
		// Reaching this would mean reusing a nonce on the next message. At
		// roughly one message per millisecond it takes half a billion years,
		// but refusing is free and silent nonce reuse is not.
		return nil, ErrCounterWrapped
	}
	e.sendCounter++

	out := make([]byte, counterSize, counterSize+len(plaintext)+e.aead.NonceSize())
	binary.BigEndian.PutUint64(out, e.sendCounter)

	sealed := e.aead.Seal(nil, nonce(e.sendDir, e.sendCounter), plaintext, aad(method, path, version))
	return append(out, sealed...), nil
}

// Open decrypts a payload, refusing replays and out-of-order counters.
//
// The counter check happens before decryption so a flood of replayed messages
// costs a comparison rather than an AEAD operation.
func (e *Envelope) Open(method, path string, version int, sealed []byte) ([]byte, error) {
	if len(sealed) < counterSize {
		return nil, ErrMalformed
	}

	counter := binary.BigEndian.Uint64(sealed[:counterSize])

	e.mu.Lock()
	defer e.mu.Unlock()

	// Strictly increasing. Equal is a replay; lower is either a replay or a
	// reordered message, and neither is something a local network request
	// should produce.
	if e.seenAny && counter <= e.recvHighWater {
		return nil, ErrReplay
	}
	if counter == 0 {
		return nil, ErrMalformed // counters start at 1
	}

	plaintext, err := e.aead.Open(nil, nonce(e.recvDir, counter), sealed[counterSize:], aad(method, path, version))
	if err != nil {
		// Do not advance the high-water mark on a failed open: an attacker
		// sending garbage with a huge counter would otherwise be able to
		// wedge the session against its legitimate peer.
		return nil, ErrAuthFailed
	}

	e.recvHighWater = counter
	e.seenAny = true
	return plaintext, nil
}

// SendCounter reports how many messages this side has sealed. Tests use it;
// nothing in the protocol depends on it.
func (e *Envelope) SendCounter() uint64 {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.sendCounter
}
